src/order/well_founded_set.lean

```diff
@@ -202,6 +207,39 @@ theorem partially_well_ordered_on.image_of_monotone_on {s : set α}
     exact (classical.some_spec (h n)).1 }
 end
 
+lemma _root_.is_antichain.finite {s : set α} {r : α → α → Prop} (ha : is_antichain r s)
+  (hp : s.partially_well_ordered_on r) :
+  s.finite :=
+begin
+  refine finite_or_infinite.resolve_right (λ hi, _),
+  obtain ⟨m, n, hmn, h⟩ := hp (λ n, hi.nat_embedding _ n) (range_subset_iff.2 $
+    λ n, (hi.nat_embedding _ n).2),
+  exact hmn.ne ((hi.nat_embedding _).injective $ subtype.val_injective $
+    ha.eq_of_related (hi.nat_embedding _ m).2 (hi.nat_embedding _ n).2 h),
+end
+
+lemma partially_well_ordered_on_iff_finite_antichains {s : set α} {r : α → α → Prop} [is_refl α r]
```

Comment:
Can we use the following lemmas to simplify this proof:
```lean
-- This could be placed just below `infinite.exists_lt_map_eq_of_maps_to`
theorem infinite.exists_lt_map_eq_of_range_subset
  {α β : Type*} [linear_order α] [_root_.infinite α] {t : set β} {f : α → β}
  (hf : range f ⊆ t) (ht : finite t) :
  ∃ x y, x < y ∧ f x = f y :=
begin
  replace hf : maps_to f (univ : set α) t, { simp [range_subset_iff.mp hf], },
  obtain ⟨x, -, y, -, hxy⟩ := (@infinite_univ α _).exists_lt_map_eq_of_maps_to hf ht,
  exact ⟨x, y, hxy⟩,
end

lemma partially_well_ordered_on_of_finite_antichain
  {s : set α} {r : α → α → Prop} [is_refl α r] (ha : is_antichain r s) (hs : s.finite) :
  s.partially_well_ordered_on r :=
begin
  intros f hf,
  suffices : ∃ m n, m < n ∧ f m = f n,
  { obtain ⟨m, n, h₁, h₂⟩ := this,
    refine ⟨m, n, h₁, _⟩,
    rw h₂,
    exact refl (f n), },
  exact infinite.exists_lt_map_eq_of_range_subset hf hs,
end
```

Reply:
I don't really see how, because `s` is most likely not finite and that doesn't give us any useful information on `t`. But nice lemmas! I incorporated them.